fabric/src/main/java/me/lucko/luckperms/fabric/mixin/ServerConfigurationNetworkHandlerMixin.java

```diff
@@ -0,0 +1,66 @@
+ *  Copyright (c) lucko (Luck) <luck@lucko.me>
+ *  Copyright (c) contributors
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.fabric.mixin;
+
+import com.llamalad7.mixinextras.sugar.Local;
+import com.mojang.authlib.GameProfile;
+import me.lucko.luckperms.fabric.event.PreOnPlayerConnectCallback;
+import me.lucko.luckperms.fabric.event.ServerConfigurationTickCallback;
+import net.minecraft.network.ClientConnection;
+import net.minecraft.network.packet.c2s.config.ReadyC2SPacket;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.network.ConnectedClientData;
+import net.minecraft.server.network.ServerCommonNetworkHandler;
+import net.minecraft.server.network.ServerConfigurationNetworkHandler;
+import net.minecraft.server.network.ServerPlayerEntity;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(ServerConfigurationNetworkHandler.class)
+public abstract class ServerConfigurationNetworkHandlerMixin extends ServerCommonNetworkHandler {
+    @Shadow @Final private GameProfile profile;
+
+    public ServerConfigurationNetworkHandlerMixin(MinecraftServer server, ClientConnection connection, ConnectedClientData clientData) {
+        super(server, connection, clientData);
+    }
+
+    @Inject(method = "onReady", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/PlayerManager;onPlayerConnect(Lnet/minecraft/network/ClientConnection;Lnet/minecraft/server/network/ServerPlayerEntity;Lnet/minecraft/server/network/ConnectedClientData;)V", shift = At.Shift.BEFORE), cancellable = true)
+    private void beforeConnecttion(ReadyC2SPacket packet, CallbackInfo ci, @Local ServerPlayerEntity player) {
```

Comment:
Is there not a way to do this with Fabric API? :(

I would prefer to now have to maintain our own event for something as critical (to LuckPerms) as the login event

Reply:
All fabric events happen too late from my testing. This is the earliest point ServerPlayerEntity instance exists and it runs just before rest of server's initialization code does, which should cover all cases (as mods are also likely to check permissions with mixin at any point within the onPlayerConnect). Fabric is just a mixin-heavy platform, so there is no real way to avoid it.